samza-core/src/main/java/org/apache/samza/task/SystemTimerScheduler.java

```diff
@@ -63,6 +63,7 @@ private SystemTimerScheduler(ScheduledExecutorService executor) {
 
     final long delay = timestamp - System.currentTimeMillis();
     final ScheduledFuture<?> scheduledFuture = executor.schedule(() -> {
+        scheduledFutures.remove(key);
```

Comment:
but , isn't there a check two lines above that checks that key is not in this map?

Reply:
Right, the removal is in the callback.

Reply:
ok, got it.